### PR TITLE
 Enhance layout optimization logic to remove duplicated layout transp…

### PR DESCRIPTION
**Pull request by yangjunpro**
Enhance layout optimization logic to remove duplicated layout transpose operation for performance improvement purpose.
    For example:
            a NCHW-to-NHWC transpose op followed by a NHWC-to-NCHW
            transpose op, under this scenario these two ops acutally can be absorbed into
            nothing, thus avoid unnecessary computation.

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc [hunk 1]
```diff
@@ -1980,10 +2014,13 @@ class DataLayoutOptimizer : GraphProcessor {
...
       node->mutable_attr()->erase("_output_shapes");
       if (IsTransposeNHWCToNCHW(node->name()) ||
+          IsUserDefinedTransposeNHWCToNCHW(node, node_map_) ||
           IsDimMapNHWCToNCHW(node->name()) ||
           IsVecPermuteNHWCToNCHW(node->name())) {
-        bool transpose_pair = IsTransposeNHWCToNCHW(node->name()) &&
-                              IsTransposeNCHWToNHWC(node->input(0));
+        bool transpose_pair = (IsTransposeNHWCToNCHW(node->name()) ||
+                               IsUserDefinedTransposeNHWCToNCHW(node, node_map_)) &&
+                              (IsTransposeNCHWToNHWC(node->input(0)) ||
```
Comment (zhangyaobit) [inline tensorflow/core/grappler/optimizers/layout_optimizer.cc]: Same here?

**Review by zhangyaobit — COMMENTED** [review]
Thank you for your contribution! This looks good overall. 

Could you comment on the use case for this PR? I'm curious why an user would have this kind of cancelable pairs in their code?

Could you add an unit test?

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc [hunk 2]
```diff
@@ -226,6 +227,39 @@ bool IsNodeType(const string& node_name, const string& type) {
   return EndWith(node_name, suffix);
 }
 
+bool IsUserDefinedTranspose(const NodeDef* node, const NodeMap* node_map,
```
[the remark this answers is not in the document]
Comment (yangjunpro) [reply tensorflow/core/grappler/optimizers/layout_optimizer.cc]: Agree. Consistency with coding convention is better.

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc [hunk 3]
```diff
@@ -226,6 +227,39 @@ bool IsNodeType(const string& node_name, const string& type) {
...
+                            const std::vector<int>& perm) {
+  if (node->op() == "Transpose" && node->input_size() == 2) {
+    const NodeDef* perm_node = node_map->GetNode(node->input(1));
+    if (!IsConstant(*perm_node)) {
+      return false;
+    }
+    const TensorProto tensor = perm_node->attr().at("value").tensor();
```
[the remark this answers is not in the document]
Comment (yangjunpro) [reply tensorflow/core/grappler/optimizers/layout_optimizer.cc]: Could you please elaborate your points a little bit?

I think the coding convention is as robust as what is doing in UpdateAttrValue().

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc [hunk 4]
```diff
@@ -226,6 +227,39 @@ bool IsNodeType(const string& node_name, const string& type) {
...
+      if (raw_values[i] != perm[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+bool IsUserDefinedTransposeNHWCToNCHW(const NodeDef* node, const NodeMap* node_map) {
```
[the remark this answers is not in the document]
Comment (yangjunpro) [reply tensorflow/core/grappler/optimizers/layout_optimizer.cc]: Replied as above.

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc [hunk 5]
```diff
@@ -1980,10 +2014,13 @@ class DataLayoutOptimizer : GraphProcessor {
       auto node = graph_->mutable_node(i);
       node->mutable_attr()->erase("_output_shapes");
       if (IsTransposeNHWCToNCHW(node->name()) ||
+          IsUserDefinedTransposeNHWCToNCHW(node, node_map_) ||
           IsDimMapNHWCToNCHW(node->name()) ||
           IsVecPermuteNHWCToNCHW(node->name())) {
-        bool transpose_pair = IsTransposeNHWCToNCHW(node->name()) &&
-                              IsTransposeNCHWToNHWC(node->input(0));
+        bool transpose_pair = (IsTransposeNHWCToNCHW(node->name()) ||
```
[the remark this answers is not in the document]
Comment (yangjunpro) [reply tensorflow/core/grappler/optimizers/layout_optimizer.cc]: The logic is necessary  for checking whether we need to remove two consecutive transform ops.

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc (hunk 1 above)
Comment (yangjunpro) [reply tensorflow/core/grappler/optimizers/layout_optimizer.cc]: Same as above reply:)

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhangyaobit) [reply tensorflow/core/grappler/optimizers/layout_optimizer.cc]: Does IsUserDefinedTransposeNHWCToNCHW already cover the functionality of IsTransposeNHWCToNCHW? If so, could you rename IsUserDefinedTransposeNHWCToNCHW to IsTransposeNHWCToNCHW, and remove the original IsTransposeNHWCToNCHW?

File: tensorflow/core/grappler/optimizers/layout_optimizer.cc [hunk 6]
```diff
@@ -226,6 +227,39 @@ bool IsNodeType(const string& node_name, const string& type) {
   return EndWith(node_name, suffix);
 }
 
+bool IsUserDefinedTranspose(const NodeDef* node, const NodeMap* node_map,
+                            const std::vector<int>& perm) {
+  if (node->op() == "Transpose" && node->input_size() == 2) {
```
Comment (zhangyaobit) [inline tensorflow/core/grappler/optimizers/layout_optimizer.cc]: If this Transpose node is a fanin of other nodes in the graph, you can't delete it. You can check this with NodeMap::GetOutputs. Could you add two test cases, one for this "not cacenllable" case (is a fanin for other nodes) and one for the cancelable case (is not fanin for other nodes)?

Comment (zhangyaobit) [reply tensorflow/core/grappler/optimizers/layout_optimizer.cc]: Tests could be added here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/grappler/layout_optimizer_test.py

**Comment by rmlarsen** [comment]
@yangjunpro any progress addressing the comments?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmlarsen** [comment]
@yangjunpro friendly poing. any progress?

**Comment by tensorflowbutler** [comment]
It has been 28 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

**Pull request closed without merging**